prov/rxm/src/rxm_ep.c

```diff
@@ -1837,103 +1892,145 @@ static ssize_t rxm_ep_trecvv(struct fid_ep *ep_fid, const struct iovec *iov,
 
-	return rxm_ep_inject_send(rxm_ep, buf, len, dest_addr, 0,
+	ret = rxm_ep_prepare_tx(rxm_ep, dest_addr, &rxm_conn);
+	if (OFI_UNLIKELY(ret))
+		return ret;
+
+	return rxm_ep_inject_send(rxm_ep, rxm_conn, buf, len, 0,
 				  rxm_ep->util_ep.inject_op_flags, tag,
 				  ofi_op_tagged);
 }
 
 static ssize_t rxm_ep_tinject_fast(struct fid_ep *ep_fid, const void *buf, size_t len,
 			      fi_addr_t dest_addr, uint64_t tag)
 {
+	int ret;
+	struct rxm_conn *rxm_conn;
 	struct rxm_ep *rxm_ep = container_of(ep_fid, struct rxm_ep,
 					     util_ep.ep_fid.fid);
 
-	return rxm_ep_inject_send_fast(rxm_ep, buf, len, dest_addr,
-				       rxm_ep->util_ep.inject_op_flags, tag,
-				       rxm_ep->tinject_tx_pkt);
+	ret = rxm_ep_prepare_tx(rxm_ep, dest_addr, &rxm_conn);
+	if (OFI_UNLIKELY(ret))
+		return ret;
+
+	return rxm_ep_tinject_send_fast(rxm_ep, rxm_conn, buf, len, tag);
 }
 
 static ssize_t rxm_ep_tsenddata(struct fid_ep *ep_fid, const void *buf, size_t len,
 				void *desc, uint64_t data, fi_addr_t dest_addr,
 				uint64_t tag, void *context)
 {
+	int ret;
+	struct rxm_conn *rxm_conn;
 	struct iovec iov = {
 		.iov_base = (void *)buf,
 		.iov_len = len,
 	};
 	struct rxm_ep *rxm_ep = container_of(ep_fid, struct rxm_ep,
 					     util_ep.ep_fid.fid);
 
-	return rxm_ep_send_common(rxm_ep, &iov, desc, 1, dest_addr, context, data,
+	ret = rxm_ep_prepare_tx(rxm_ep, dest_addr, &rxm_conn);
+	if (OFI_UNLIKELY(ret))
+		return ret;
+
+	return rxm_ep_send_common(rxm_ep, rxm_conn, &iov, desc, 1, context, data,
 				  rxm_ep_tx_flags(rxm_ep) | FI_REMOTE_CQ_DATA,
-				  tag, ofi_op_tagged, rxm_ep->tinject_tx_pkt);
+				  tag, ofi_op_tagged, rxm_conn->tinject_data_pkt);
 }
 
 static ssize_t rxm_ep_tinjectdata(struct fid_ep *ep_fid, const void *buf, size_t len,
 				  uint64_t data, fi_addr_t dest_addr, uint64_t tag)
 {
+	int ret;
+	struct rxm_conn *rxm_conn;
 	struct rxm_ep *rxm_ep = container_of(ep_fid, struct rxm_ep,
 					     util_ep.ep_fid.fid);
 
-	return rxm_ep_inject_send(rxm_ep, buf, len, dest_addr, data,
+	ret = rxm_ep_prepare_tx(rxm_ep, dest_addr, &rxm_conn);
+	if (OFI_UNLIKELY(ret))
+		return ret;
+
+	return rxm_ep_inject_send(rxm_ep, rxm_conn, buf, len, data,
 				  rxm_ep->util_ep.inject_op_flags |
-				  FI_REMOTE_CQ_DATA,
-				  tag, ofi_op_tagged);
+				  FI_REMOTE_CQ_DATA, tag, ofi_op_tagged);
 }
 
 static ssize_t rxm_ep_tinjectdata_fast(struct fid_ep *ep_fid, const void *buf, size_t len,
 				       uint64_t data, fi_addr_t dest_addr, uint64_t tag)
 {
+	int ret;
+	struct rxm_conn *rxm_conn;
 	struct rxm_ep *rxm_ep = container_of(ep_fid, struct rxm_ep,
 					     util_ep.ep_fid.fid);
 
-	return rxm_ep_inject_data_send_fast(rxm_ep, buf, len, dest_addr, data,
-					    rxm_ep->util_ep.inject_op_flags |
-					    FI_REMOTE_CQ_DATA,
-					    tag, rxm_ep->tinject_tx_pkt);
+	ret = rxm_ep_prepare_tx(rxm_ep, dest_addr, &rxm_conn);
+	if (OFI_UNLIKELY(ret))
+		return ret;
+
+	return rxm_ep_tinject_data_send_fast(rxm_ep, rxm_conn, buf, len, data, tag);
```

Comment:
Since we already have access to `rxm_conn->tinject_data_pkt` in this function, we could just set the data field in header here and pass `tinject_data_pkt` as argument to a common *send_fast function which would set the remaining fields. Same comment for non-tagged functions.

Reply:
Goo catch! Thank you. Done